Framework/Algorithms/src/FindPeaksConvolve.cpp

```diff
@@ -0,0 +1,376 @@
+  Tensor1D kernel(kernelSize);
+  for (size_t i{0}; i < kernelSize; i++) {
+    kernel.data()[i] = 1.0 / static_cast<double>(kernelSize);
+  }
+  return kernel;
+}
+
+size_t FindPeaksConvolve::getKernelBinCount(const HistogramData::HistogramX *xData) const {
+  // What is the minimum number of data points this works for - add validation
+  const double peakExtent = getProperty("EstimatedPeakExtent");
+  const int peakExtentNBins = getProperty("EstimatedPeakExtentNBins");
+
+  if (peakExtent != EMPTY_DBL()) {
+    const double x1{xData->rawData()[static_cast<size_t>(std::floor((xData->size() - 1) / 2))]};
+    const double x2{xData->rawData()[static_cast<size_t>(std::floor((xData->size() - 1) / 2)) + 1]};
+    return static_cast<size_t>(std::floor(peakExtent * 2 / (x2 - x1)));
+  } else {
+    return static_cast<size_t>(peakExtentNBins);
+  }
+}
+
+Eigen::VectorXd FindPeaksConvolve::centreBinsXData(const HistogramData::HistogramX *xData) {
+  Eigen::VectorXd xDataVec{0.5 * (EigenMap_const(&xData->front(), xData->size() - 1) +
+                                  EigenMap_const(&xData->front() + 1, xData->size() - 1))};
+  return xDataVec;
+}
+
+void FindPeaksConvolve::extractPeaks(const size_t dataIndex, const Tensor1D &iOverSigma, const EigenMap_const &xData,
+                                     const TensorMap_const &yData, const size_t peakExtentBinNumber) {
+  auto filteredData = filterDataForSignificantPoints(iOverSigma);
+  int dataPointCount{0};
+  std::pair<int, double> dataRegionMax{0, 0.0};
+  std::vector<FindPeaksConvolve::PeakResult> peakCentres;
+  for (const auto &dataPoint : filteredData) {
+    if (dataPoint.second != 0) {
+      if (dataPointCount == 0) {
+        dataRegionMax = dataPoint;
+      } else {
+        if (dataPoint.second > dataRegionMax.second) {
+          dataRegionMax = dataPoint;
+        }
+      }
+      dataPointCount++;
+    } else {
+      if (dataPointCount >= 2) {
+        size_t rawPeakIndex{findPeakInRawData(dataRegionMax.first, yData, peakExtentBinNumber)};
+        peakCentres.push_back(
+            FindPeaksConvolve::PeakResult{xData[rawPeakIndex], yData.data()[rawPeakIndex], dataRegionMax.second});
+      }
+      if (dataPointCount > 0) {
+        dataPointCount = 0;
+        dataRegionMax = std::make_pair(0, 0.0);
+      }
+    }
+  }
+  storePeakResults(dataIndex, peakCentres);
+}
+
+void FindPeaksConvolve::storePeakResults(const size_t dataIndex,
+                                         std::vector<FindPeaksConvolve::PeakResult> &peakCentres) {
+  const size_t peakCount = peakCentres.size();
+  if (peakCount) {
+    if (peakCount > m_maxPeakCount) {
+      m_maxPeakCount = peakCount;
+    }
+    m_peakResults[dataIndex] = std::move(peakCentres);
+  }
+}
+
+std::vector<std::pair<const int, const double>>
+FindPeaksConvolve::filterDataForSignificantPoints(const Tensor1D &iOverSigma) {
+  std::vector<std::pair<const int, const double>> closedData;
```

Comment:
Is this not making quite an expensive copy of I/sigma from convolution? Unless I've misunderstood ], it looks to me like a vector of  index, I/sigma pairs (albeit with some I/sigma some zeroed outside of the peak regions) - surely we don't need to store the index?

I would say there are two options here:

1. Store pairs of `istart, iend` here
2. Make a store a vector of I'sigma with some elements zeroed

My preference would be for (1) but it depends on how it fits in with everything else!